amm-executor/src/Spectrum/Executor/Backlog/Service.hs

```diff
@@ -1,60 +1,158 @@
+        modifyIORef pendingPQ (PQ.insert $ mkWeightedOrderWithTimestamp order timestamp)
+    , suspend = \(SuspendedOrder order timestamp) -> do
+        modifyIORef suspendedPQ (PQ.insert $ mkWeightedOrderWithTimestamp order timestamp)
+        isExist $ BacklogOrder timestamp order
+    , checkLater = \(InProgressOrder order timestamp) -> do
+        modifyIORef toRevisitQ (mkWeightedOrderWithTimestamp order timestamp Seq.<|)
+        isExist $ BacklogOrder timestamp order
+    , tryAcquire = do
+        _ <- refreshQueues config store pendingPQ suspendedPQ toRevisitQ
+        maxWeightedOrderM <- getMaxWeightedOrder' pendingPQ suspendedPQ
+        case maxWeightedOrderM of
+          Just (WeightedOrderWithTimestamp oId _ _) -> get oId <&> fmap backlogOrder
+          Nothing -> pure Nothing
+    , drop = \(ExecutedOrder oId) -> do
+        _ <- modifyIORef pendingPQ   (PQ.filter (\(WeightedOrderWithTimestamp orderId _ _) -> orderId /= oId))
+        _ <- modifyIORef suspendedPQ (PQ.filter (\(WeightedOrderWithTimestamp orderId _ _) -> orderId /= oId))
+        _ <- modifyIORef toRevisitQ  (Seq.filter (\(WeightedOrderWithTimestamp orderId _ _) -> orderId /= oId))
+        drop oId
+    }
 
-  (_, db) <- Rocks.openBracket storePath
-              Rocks.defaultOptions
-                { Rocks.createIfMissing = createIfMissing
-                }
+refreshQueues
+  :: (MonadIO m)
+  => BacklogServiceConfig
+  -> BacklogStore m
+  -> IORef (PQ.MaxQueue WeightedOrderWithTimestamp)
+  -> IORef (PQ.MaxQueue WeightedOrderWithTimestamp)
+  -> IORef (Seq.Seq WeightedOrderWithTimestamp)
+  -> m ()
+refreshQueues BacklogServiceConfig{..} BacklogStore{..} pendingQueueRef suspendedQueueRef toRevisitSeqRef = do
+  currentTime <- getCurrentTime
+  revisited2process <-
+    atomicModifyIORef'
+      toRevisitSeqRef
+      (Seq.partition (\(WeightedOrderWithTimestamp _ _ oTime) -> diffUTCTime currentTime oTime < orderExecTime))
+  suspended2drop   <-
+    atomicModifyIORef'
+      suspendedQueueRef
+      (PQ.partition (\(WeightedOrderWithTimestamp _ _ oTime) -> diffUTCTime currentTime oTime < orderLifetime))
+  _ <- (\order@(WeightedOrderWithTimestamp oId _ oTime) ->
+          if diffUTCTime currentTime oTime < orderLifetime
+            then modifyIORef' pendingQueueRef (PQ.insert order)
+            else drop oId
+       ) `traverse` revisited2process
+  _ <- (\(WeightedOrderWithTimestamp oId _ _) -> drop oId) `traverse` PQ.toList suspended2drop
+  pure ()
+
+getMaxWeightedOrder'
+  :: (MonadIO m)
+  => IORef (PQ.MaxQueue WeightedOrderWithTimestamp)
+  -> IORef (PQ.MaxQueue WeightedOrderWithTimestamp)
+  -> m (Maybe WeightedOrderWithTimestamp)
+getMaxWeightedOrder' pendingQueueRef suspendedQueueRef = do
+  pendingQueue   <- readIORef pendingQueueRef
+  suspendedQueue <- readIORef suspendedQueueRef
+  let
+    maxPendingM   = PQ.getMax pendingQueue
+    maxSuspendedM = PQ.getMax suspendedQueue
+  case (maxPendingM, maxSuspendedM) of
```

Comment:
It seems we are missing processing of orders from `toRevisit` queue. Those should be checked in the beginning of selection:
```
select = do
  ordRev <- Seq.head toRevisitQueue
  if ordRev.checkAt <= currentTimeMillis then return ordRev else getMaxWeightedOrder'
```

Reply:
nope, check https://github.com/ergolabs/cardano-dex-backend/blob/63de8c7b18f37e1d5fb4106025480018ee70d84d/amm-executor/src/Spectrum/Executor/Backlog/Service.hs#L57

Reply:
Ok, those orders should be processed passively as well as suspended ones

Reply:
toRevisit is seq. It wouldn't be efficient to use such approach

Reply:
So you think calling `PQ.partition`, `Seq.partition`, `PQ.toList` each time is more efficient?

Reply:
https://hackage.haskell.org/package/containers-0.6.0.1/docs/Data-Sequence.html